container/state.go

```diff
@@ -160,64 +165,78 @@ func IsValidStateString(s string) bool {
+// WaitCondition is an enum type for different states to wait for.
+type WaitCondition int
+
+// Possible WaitCondition Values.
+//
+// WaitConditionNotRunning (default) is used to wait for any of the non-running
+// states: "created", "exited", "dead", "removing", or "removed".
+//
+// WaitConditionNextExit is used to wait for the next time the state changes
+// to a non-running state. If the state is currently "created" or "exited",
+// this would cause Wait() to block until either the container runs and exits
+// or is removed.
+//
+// WaitConditionRemoved is used to wait for the container to be removed.
+const (
+	WaitConditionNotRunning WaitCondition = iota
+	WaitConditionNextExit
+	WaitConditionRemoved
+)
+
+// Wait waits until the continer is in a certain state indicated by the given
+// condition. A context can be used for cancelling the request or controlling
+// timeouts. Wait must be called without holding the state lock. Returns a
+// channel which can be used to receive the result. If the container exited on
+// its own, the result's Err() method wil be nil and its ExitCode() method will
+// return the conatiners exit code, otherwise, the results Err() method will
+// return an error indicating why the wait operation failed.
+func (s *State) Wait(ctx context.Context, condition WaitCondition) <-chan *StateStatus {
+	s.Lock()
+	defer s.Unlock()
 
-// WaitStop waits until state is stopped. If state already stopped it returns
-// immediately. If you want wait forever you must supply negative timeout.
-// Returns exit code, that was passed to SetStopped
-func (s *State) WaitStop(timeout time.Duration) (int, error) {
-	ctx := context.Background()
-	if timeout >= 0 {
-		var cancel func()
-		ctx, cancel = context.WithTimeout(ctx, timeout)
-		defer cancel()
+	if condition == WaitConditionNotRunning && !s.Running {
+		// Return the current status.
+		result := newStateStatus(s.ExitCode(), s.Err())
+		// Buffer so we can put it in the channel now.
+		resultC := make(chan *StateStatus, 1)
+		resultC <- result
+		return resultC
 	}
-	if err := s.WaitWithContext(ctx); err != nil {
-		if status, ok := err.(*StateStatus); ok {
-			return status.ExitCode(), nil
-		}
-		return -1, err
+
+	// If we are waiting only for removal, the waitStop channel should
+	// remain nil and block forever.
+	var waitStop chan struct{}
+	if condition < WaitConditionRemoved {
+		waitStop = s.waitStop
 	}
-	return 0, nil
-}
 
-// WaitWithContext waits for the container to stop. Optional context can be
-// passed for canceling the request.
-func (s *State) WaitWithContext(ctx context.Context) error {
-	s.Lock()
-	if !s.Running {
-		state := newStateStatus(s.ExitCode(), s.Error())
-		defer s.Unlock()
-		if state.ExitCode() == 0 {
-			return nil
+	// Always wait for removal, just in case the container gets removed
+	// while it is still in a "created" state, in which case it is never
+	// actually stopped.
+	waitRemove := s.waitRemove
+
+	resultC := make(chan *StateStatus)
```

Comment:
It doesn't need a buffer in this case since blocking on it doesn't prevent this function from returning. It's buffered in the other case (when the condition and container is `not-running`) because we can avoid needing to spawn a goroutine to put it in the channel.